runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactRetrievalService.java

```diff
@@ -0,0 +1,200 @@
+    try {
+      ArtifactApi.ProxyManifest proxyManifest = MANIFEST_CACHE.get(token);
+      ArtifactApi.GetManifestResponse response =
+          ArtifactApi.GetManifestResponse.newBuilder()
+              .setManifest(proxyManifest.getManifest())
+              .build();
+      LOG.info(
+          "GetManifest for {} -> {} artifacts",
+          token,
+          proxyManifest.getManifest().getArtifactCount());
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      LOG.info("GetManifest for {} failed", token, e);
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void getArtifact(
+      ArtifactApi.GetArtifactRequest request,
+      StreamObserver<ArtifactApi.ArtifactChunk> responseObserver) {
+    LOG.info("GetArtifact {}", request);
+    String name = request.getName();
+    try {
+      ArtifactApi.ProxyManifest proxyManifest = MANIFEST_CACHE.get(request.getRetrievalToken());
+      // look for file at URI specified by proxy manifest location
+      ArtifactApi.ProxyManifest.Location location =
+          proxyManifest
+              .getLocationList()
+              .stream()
+              .filter(loc -> loc.getName().equals(name))
+              .findFirst()
+              .orElseThrow(
+                  () ->
+                      new FileNotFoundException(
+                          String.format("Artifact location not found in manifest: %s", name)));
+
+      List<ArtifactMetadata> existingArtifacts = proxyManifest.getManifest().getArtifactList();
+      ArtifactMetadata metadata =
+          existingArtifacts
+              .stream()
+              .filter(meta -> meta.getName().equals(name))
+              .findFirst()
+              .orElseThrow(
+                  () ->
+                      new FileNotFoundException(
+                          String.format("Artifact metadata not found in manifest: %s", name)));
+
+      ResourceId artifactResourceId =
+          FileSystems.matchNewResource(location.getUri(), false /* is directory */);
+      LOG.info("Artifact {} located in {}", name, artifactResourceId);
+      Hasher hasher = Hashing.md5().newHasher();
+      byte[] data = new byte[ARTIFACT_CHUNK_SIZE_BYTES];
+      try (InputStream stream = Channels.newInputStream(FileSystems.open(artifactResourceId))) {
+        int len;
+        while ((len = stream.read(data)) != -1) {
+          hasher.putBytes(data, 0, len);
+          responseObserver.onNext(
+              ArtifactApi.ArtifactChunk.newBuilder()
+                  .setData(ByteString.copyFrom(data, 0, len))
+                  .build());
+        }
+      }
+      if (metadata.getMd5() != null && !metadata.getMd5().isEmpty()) {
+        ByteString expected = ByteString.copyFrom(Base64.decodeBase64(metadata.getMd5()));
+        ByteString actual = ByteString.copyFrom(hasher.hash().asBytes());
+        if (!actual.equals(expected)) {
+          throw new IllegalStateException(
```

Comment:
Should we call `responseObserver.onError`

Reply:
It gets called below in case of any exception.